data_models.py

```diff
@@ -125,64 +126,131 @@ def nmf(self, n_components, transform_df=False, transform_test_df=False,
                                         columns=self.df.columns,
                                         index=colnames)
         if transform_df:
-            out_df = self.nmf_fit.transform(transform_df)
+            out_df = self.nmf_fit.transform(self.df)
             return out_df
 
         if transform_test_df:
             self.nmf_test_df = self.nmf_fit.transform(self.test_df)
 
 
-    def combine_models(self, include_labels=False, include_raw=False,
-                       test_set=False):
-        """
-        Merge z matrices together across algorithms
+    def plier(self, n_components, transform_df=False, transform_test_df=False,
+              shuffled=False, seed=1):
+        import os
+        import subprocess
+        import tempfile
+        plier_output_dir = os.path.join(cfg.data_dir, 'plier_output')
+        if not os.path.exists(plier_output_dir):
+            os.makedirs(plier_output_dir)
+        output_prefix = os.path.join(plier_output_dir, 'plier_k{}_s{}'.format(
+                                       n_components, seed))
+        if shuffled:
+            output_prefix += '_shuffled'
+        output_data = output_prefix + '_z.tsv'
+        output_weights = output_prefix + '_b.tsv'
+        output_l2 = output_prefix + '_l2.tsv'
+
+        if (not os.path.exists(output_data) or
+            not os.path.exists(output_weights)):
+            tf = tempfile.NamedTemporaryFile(mode='w')
+            self.df.to_csv(tf, sep='\t')
+            args = [
+                'Rscript',
+                os.path.join(cfg.scripts_dir, 'run_plier.R'),
+                '--data', tf.name,
+                '--k', str(n_components),
+                '--seed', str(seed),
+                '--output_prefix', output_prefix
+            ]
+            subprocess.check_call(args)
```

Comment:
I'm not that familiar with `subprocess`, but is it true that the following line (`tf.close()`) will only be run once the `subprocess` call has finished?

Reply:
It should be true: `subprocess.check_call()` blocks execution until the child process has returned. There are other `subprocess` functions that are non-blocking, though, so it's a good thing to keep in mind. See here for more detail: https://stackoverflow.com/questions/21936597/blocking-and-non-blocking-subprocess-calls

I wrote a short script to test this, and it does seem to be the case.